pkg/sql/backfill.go

```diff
@@ -1551,10 +1552,38 @@ func ValidateCheckConstraint(
 		semaCtx.TableNameResolver = resolver
 		defer func() { descriptors.ReleaseAll(ctx) }()
 
-		return ie.WithSyntheticDescriptors([]catalog.Descriptor{tableDesc}, func() error {
-			return validateCheckExpr(ctx, &semaCtx, txn, sessionData, checkConstraint.GetExpr(),
-				tableDesc.(*tabledesc.Mutable), ie, indexIDForValidation)
-		})
+		if ck := constraint.AsCheck(); ck != nil {
+			return ie.WithSyntheticDescriptors(
+				[]catalog.Descriptor{tableDesc},
+				func() error {
+					return validateCheckExpr(ctx, &semaCtx, txn, sessionData, ck.GetExpr(),
+						tableDesc.(*tabledesc.Mutable), ie, indexIDForValidation)
+				},
+			)
+		} else if fk := constraint.AsForeignKey(); fk != nil {
+			targetTable, err := descriptors.GetImmutableTableByID(ctx, txn, fk.GetReferencedTableID(), tree.ObjectLookupFlags{
+				CommonLookupFlags: tree.CommonLookupFlags{
+					AvoidLeased:    true,
+					IncludeDropped: true,
+					IncludeOffline: true,
+				},
+			})
+			if err != nil {
+				return err
+			}
+			if targetTable.GetID() == tableDesc.GetID() {
+				targetTable = tableDesc
+			}
+			return ie.WithSyntheticDescriptors(
+				[]catalog.Descriptor{tableDesc},
+				func() error {
+					return validateForeignKey(ctx, tableDesc.(*tabledesc.Mutable), targetTable, fk.ForeignKeyDesc(),
+						indexIDForValidation, txn, ie)
+				},
+			)
+		} else {
+			return errors.AssertionFailedf("validation of unsupported constraint type")
+		}
```

Comment:
very minor nit: is it possible to `switch constraint.(type)` here instead? I'm not actually sure, in any case these if-elses feel kinda clunky.